controllers/integration/integration_adapter.go

```diff
@@ -0,0 +1,310 @@
+		if a.compareApplicationSnapshots(expectedApplicationSnapshot, &foundApplicationSnapshot) {
+			return &foundApplicationSnapshot, nil
+		}
+	}
+	return nil, nil
+}
+
+// getAllApplicationComponents loads from the cluster all Components associated with the given Application.
+// If the Application doesn't have any Components or this is not found in the cluster, an error will be returned.
+func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Application) (*[]hasv1alpha1.Component, error) {
+	applicationComponents := &hasv1alpha1.ComponentList{}
+	opts := []client.ListOption{
+		client.InNamespace(application.Namespace),
+		client.MatchingFields{"spec.application": application.Name},
+	}
+
+	err := a.client.List(a.context, applicationComponents, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationComponents.Items, nil
+}
+
+// getAllApplicationReleaseLinks returns the ReleaseLinks used by the application being processed. If matching
+// ReleaseLinks are not found, an error will be returned.
+func (a *Adapter) getAllApplicationReleaseLinks() (*[]releasev1alpha1.ReleaseLink, error) {
+	releaseLinks := &releasev1alpha1.ReleaseLinkList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingFields{"spec.application": a.application.Name},
+	}
+
+	err := a.client.List(a.context, releaseLinks, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releaseLinks.Items, nil
+}
+
+// getApplicationSnapshot returns the all ApplicationSnapshots in the Application's namespace nil if it's not found.
+// In the case the List operation fails, an error will be returned.
+func (a *Adapter) getAllApplicationSnapshotReleases(applicationSnapshot *releasev1alpha1.ApplicationSnapshot) (*[]releasev1alpha1.Release, error) {
+	releases := &releasev1alpha1.ReleaseList{}
+	opts := []client.ListOption{
+		client.InNamespace(applicationSnapshot.Namespace),
+		client.MatchingFields{"spec.applicationSnapshot": applicationSnapshot.Name},
+	}
+
+	err := a.client.List(a.context, releases, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releases.Items, nil
+}
+
+// createApplicationSnapshotForPipelineRun creates the ApplicationSnapshot for a given PipelineRun
+func (a *Adapter) createApplicationSnapshotForPipelineRun() (*releasev1alpha1.ApplicationSnapshot, error) {
+	applicationComponents, err := a.getAllApplicationComponents(a.application)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get all Application Components for Application %s", a.application.Name)
+	}
+	var images []releasev1alpha1.Image
+
+	for _, applicationComponent := range *applicationComponents {
+		pullSpec := applicationComponent.Status.ContainerImage
+		if applicationComponent.Name == a.component.Name {
+			var err error
+			pullSpec, err = tekton.GetOutputImage(a.pipelineRun)
+			if err != nil {
+				return nil, err
+			}
+		}
+		image := releasev1alpha1.Image{
```

Comment:
You don't really need to declare a variable for this simple block. The following looks better in my opinion:
```
images = append(images, releasev1alpha1.Image{
		Component: applicationComponent.Name,
		PullSpec:  pullSpec,
	}
}
```